utils/src/vector.c

```diff
@@ -0,0 +1,152 @@
+ *           allocation fails
+ */
+vector_t* init_vector_of_size(size_t item_size, size_t init_size) {
+    if (item_size == 0) {
+        return NULL;
+    }
+
+    vector_t* vec = malloc(sizeof(vector_t));
+    if (vec == NULL) {
+      return NULL;
+    }
+
+    vec->item_size = item_size;
+    vec->used = 0;
+    size_t true_init_size = roundup(init_size);
+    vec->capacity = true_init_size;
+
+    /* 
+     * If they requested 0 bytes, no need to malloc any memory 
+     * Realloc will act like malloc if the given pointer is null
+     */
+    if (init_size > 0) {
+        vec->contents = malloc(sizeof(char) * true_init_size * item_size);
+
+        /* Check for memory allocation failure */
+        if (vec->contents == NULL) {
+            return NULL;
+        }
+    } else {
+        vec->contents = NULL;
+    }
+
+    return vec;
+}
+
+/*!
+ * @brief     Initializes a new vector of size 0
+ *
+ * @param     item_size Size of the items to be stored in the vector in bytes,
+ *            must be greater than 0
+ *
+ * @return    A pointer to the new vector, NULL on allocation failure or when
+ *            item_size is 0
+ */
+vector_t* init_vector(size_t item_size) {
+    return init_vector_of_size(item_size, 0);
+}
+
+/*!
+ * @brief	  Extends the vector by a given amount
+ *
+ * @param[in]     vec Vector to be extended
+ *
+ * @param         change Amount to grow the vector by
+ *
+ * @return        Returns 1 on success or 0 on failure
+ */
+static int extend_vector(vector_t* vec, size_t change) {
+    if (vec->used + change <= vec->capacity) {
+        return 1;
+    }
+
+    size_t n_items = roundup(vec->used + change);
+    char* new_buffer = realloc(vec->contents, n_items * vec->item_size);
+    if (new_buffer == NULL) {
+        return 0;
+    }
+
+    vec->contents = new_buffer;
+    vec->capacity = n_items;
+    return 1;
+}
+
+/*!
+ * @brief      Pushes a single item onto the back of the vector
+ *
+ * @param[in]  vec Vector to be expanded
+ *
+ * @param[in]  item Pointer to the item to be deep copied into the vector
+ *
+ * @return     1 on success, or 0 on failure
+ */
+int push_vector(vector_t* vec, void* item) {
+    if (!extend_vector(vec, 1)) {
+        return 0;
+    }
+    memcpy(&vec->contents[vec->used * vec->item_size], item, vec->item_size);
+    vec->used += 1;
+    return 1;
+}
+
+/*!
+ * @brief      Indexes the vector and returns a pointer to the item
+ *
+ * @param[in]  vec Vector to be indexed
+ *
+ * @param      index Position of the item in the vector
+ *
+ * @return     A pointer to the item, NULL if index is invalid
+ */
+void* index_vector(vector_t* vec, size_t index) {
+    if (index >= vec->used)
+        return NULL;
+
+    return (void*)(&vec->contents[index * vec->item_size]);
+}
+
+/*!
+ * @brief     Destroys and frees the memory created by make_vector
+ *
+ * @param[in] vec Vector to be destroyed
+ */
+void free_vector(vector_t* vec) {
```

Comment:
Looks like the way you are using a vector in the program, you are not doing pass-by-reference here.
The parameter should be `vector_t ** vec` to indicate that you are passing by reference.

Reply:
This seems like pass by reference?  Also, passing a with 2 levels of indirection seems silly, as the only benefit is that you can set the 2nd pointer to NULL.  Its also extremely minimal, but 2 memory accesses is slower than 1,

Reply:
So in your tests, you initialize for example, a vector * test_vector, and then you call free_vector(test_vector) but because the parameter is vector * instead of vector **, the pointer is being passed by value, so you would not be able to set the freed pointer to NULL unless you did a pass by reference: https://stackoverflow.com/questions/6054271/freeing-pointers-from-inside-other-functions-in-c

Reply:
So we would want to set the pointer to NULL after free

Reply:
Hmm.  Passing with a double pointer seems obtuse.  Most libraries that have functions that free objects use just one pointer and leave the programmer to decide if they want to set the pointer to NULL.

Reply:
If passing a double pointer seems icky, then maybe we want to not use `vector_t *` and instead initialize vectors with `vector_t`. This way we can perform a pass by reference when we want to free the vector contents...what do you think?